qdl.c

```diff
@@ -110,6 +112,306 @@ static enum qdl_storage_type decode_storage(const char *storage)
 	exit(1);
 }
 
+#define CPIO_MAGIC "070701"
+struct cpio_newc_header {
+	char c_magic[6];       /* "070701" */
+	char c_ino[8];
+	char c_mode[8];
+	char c_uid[8];
+	char c_gid[8];
+	char c_nlink[8];
+	char c_mtime[8];
+	char c_filesize[8];
+	char c_devmajor[8];
+	char c_devminor[8];
+	char c_rdevmajor[8];
+	char c_rdevminor[8];
+	char c_namesize[8];
+	char c_check[8];
+};
+
+uint32_t parse_ascii_hex32(const char *s)
+{
+	uint32_t x = 0;
+
+	for (int i = 0; i < 8; i++) {
+		if (!isxdigit(s[i]))
+			err(1, "non-hex-digit found in archive header");
+
+		if (s[i] <= '9')
+			x = (x << 4) | (s[i] - '0');
+		else
+			x = (x << 4) | (10 + (s[i] | 32) - 'a');
+	}
+
+	return x;
+}
+
+/**
+ * decode_programmer_archive() - Attempt to decode a programmer CPIO archive
+ * @images: List of Sahara images, with @images[0] populated
+ *
+ * The single blob provided in @images[0] might be a CPIO archive containing
+ * Sahara images, in files with names in the format "<id>:<filename>". Load
+ * each such Sahara image into the relevant spot in the @images array.
+ *
+ * The original blob (in @images[0]) is freed once it has been consumed.
+ *
+ * Returns: 0 if no archive was found, 1 if archive was decoded, -1 on error
+ */
+static int decode_programmer_archive(struct sahara_image *images)
+{
+	struct sahara_image *blob = &images[0];
+	struct cpio_newc_header *hdr;
+	size_t filesize;
+	size_t namesize;
+	char name[128];
+	char *save;
+	char *tok;
+	void *ptr = blob->ptr;
+	void *end = blob->ptr + blob->len;
+	long id;
+
+	if (blob->len < sizeof(*hdr) || memcmp(ptr, CPIO_MAGIC, 6))
+		return 0;
+
+	for (;;) {
+		if (ptr + sizeof(*hdr) > end) {
+			fprintf(stderr, "programmer archive is truncated\n");
+			return -1;
```

Comment:
Not related to this PR, but I noticed that in some source files we use definitions from errno.h, while in others we use magic numbers for return statements. Maybe we should adopt a consistent approach using symbolic error names (e.g., -EINVAL) and enforce this in all new PRs.